sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +397,37 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+                missing_index = np.where(mask.any(1))[0]
+                for i, row in zip(missing_index, X[missing_index]):
+                    col_index = np.where(np.logical_not(np.isnan(row)))[0]
+                    impute_index = np.where(np.isnan(row))[0]
+                    neigh = NearestNeighbors(self.kneighbor)
```

Comment:
We could use a different metric using `NearestNeighbors`.

If only euclidean and the matrices involved aren't too large for memory, one could do something tricky like:
```python
D2 = (X[missing_index, np.newaxis] - X) ** 2  # componentwise squared differences between queries and full dataset
D2[np.isnan(D2)] = 0  # missing features from query contribute 0 to squared distance
missing_row, missing_col = np.where(np.isnan(X))
D2[:, missing_row, missing_col] = np.nan #  but missing features from index contribute NaN
sqdist = D2.sum(axis=2)
sqdist[np.arange(len(missing_index)), missing_index] = np.inf  # avoid returning query point
ind = np.argsort(sqdist, axis=1)[:, :n_neighbors]
means = np.mean(X[ind], axis=1)
out = X.copy()
out[missing_row, missing_col] = means[np.where(np.isnan(X[missing_index]))[0], missing_col]
```
(Perhaps I should have left that for you to do, but it was too much fun to miss.)